### PR TITLE
Improve performance of JSON encoding 

### PR DESCRIPTION
**Pull request by qwwdfsad**
Performance improvements, the same machine, Java 11:

```
CitmBenchmark.encodeCitm                               +41%
JacksonComparisonBenchmark.kotlinSmallToString         +70% 
JacksonComparisonBenchmark.kotlinToString              +339% (I've triple-checked that!)
JacksonComparisonBenchmark.kotlinToStringWithEscapes   +86%
TwitterBenchmark.encodeTwitter                         +71%
TwitterFeedBenchmark.encodeTwitter                     +101%
```

### CODE REVIEW

**Comment by qwwdfsad** [comment]
I suppose it's not really clear where all these improvements come from, so feel free to ask for additional comments, meanwhile, I'll try to document all my finding separately

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            val ch = string[i].toInt()
+            // Do we have unescaped symbols?
+            var marker: Char = 0.toChar()
+            if (ch < ESCAPE_MARKERS.size && ESCAPE_MARKERS[ch].also { marker = it } != 0.toChar()) {
+                if (marker != 0.toChar()) {
+                    array[sz] = '\\'
+                    array[sz + 1] = marker
+                    sz += 2
+                } else {
+                    val escapedString = ESCAPE_STRINGS[ch]!!
```
Comment (sandwwraith) [inline formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:60]: I don't get it. How we can visit this branch if marker seems to be always non-zero? (Given that this all is under `ESCAPE_MARKERS[ch].also { marker = it } != 0.toChar()` branch).

Also, how we distinguish regular chars (say 'A', which is 0x41) from 0..0x1f chars? They all have `0` value in `ESCAPE_MARKERS` array.

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        return String(array, 0, size)
+    }
+
+    private fun ensureAdditionalCapacity(expected: Int) {
+        ensureTotalCapacity(size + expected)
+    }
+
+    private fun ensureTotalCapacity(newSize: Int) {
+        if (array.size <= newSize) {
+            array = array.copyOf(newSize.coerceAtLeast(size * 2))
```
Comment (sandwwraith) [inline formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:83]: Here we can exceed max chars size. If application regularly sends large JSONs, it would still suffer from allocations. Maybe we can make sure that at least one char array always stays in pool? Or recommendation for such applications would be to increase max pool size? Then we should document this.

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/CharArrayPool.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package kotlinx.serialization.json.internal
+
+import java.util.concurrent.*
+
+internal object CharArrayPool {
+    private val arrays = ArrayDeque<CharArray>()
+    private var charsTotal = 0
+    private val MAX_CHARS_IN_POOL = runCatching {
+        System.getProperty("kotlinx.serialization.json.pool.size").toIntOrNull()
```
Comment (sandwwraith) [inline formats/json/jvmMain/src/kotlinx/serialization/json/internal/CharArrayPool.kt:9]: This property should be documented somewhere. Maybe not in the .md files, but at least in the JSON documentation.

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt (hunk 1 above)
Comment (sandwwraith) [reply formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:60]: I feel that you should tests for some corner cases, benchmarks are not enough. For example, test with vertical tab (0x0b) or something and make sure behavior stays the same.

**Review by sandwwraith — CHANGES_REQUESTED** [review]

File: formats/json/commonMain/src/kotlinx/serialization/json/internal/StringOps.kt [hunk 4]
```diff
@@ -34,21 +30,34 @@ private val ESCAPE_CHARS: Array<String?> = arrayOfNulls<String>(128).apply {
...
+}
+
 internal fun StringBuilder.printQuoted(value: String) {
     append(STRING)
     var lastPos = 0
-    val length = value.length
-    for (i in 0 until length) {
+    for (i in value.indices) {
         val c = value[i].toInt()
-        // Do not replace this constant with C2ESC_MAX (which is smaller than ESCAPE_CHARS size),
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply formats/json/commonMain/src/kotlinx/serialization/json/internal/StringOps.kt:43]: We don't, actually 93..128 check contains a lot of lower-case ASCII symbols and previously it triggered a bunch of redundant work

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/CharArrayPool.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    }.getOrNull() ?: 1024 * 1024 // 2 MB seems to be a reasonable contraint
+
+    public fun take(): CharArray {
+        val candidate = synchronized(this) {
+            arrays.lastOrNull()?.also { charsTotal -= it.size }
+        }
+        return candidate ?: CharArray(128)
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply formats/json/jvmMain/src/kotlinx/serialization/json/internal/CharArrayPool.kt:16]: More or less. I left a comment

>It helps eliminating new allocations in StringBuilder creation

Not really, otherwise we could have cached `StringBuilder` instances and invoke `setLength(0)`. Regarding this, I'll leave a comment in `JsonStringBuilder`

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+/*
+ * Optimized version of StringBuilder that is specific to JSON-encoding
+ */
+internal actual class JsonStringBuilder {
+    private var array = CharArrayPool.take()
+    private var size = 0
+
+    actual fun append(value: Long) {
+        ensureAdditionalCapacity(20) // Long length
+        append(value.toString())
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:12]: Good catch, leftover from my handrolled optimization

File: formats/json/nativeMain/src/kotlinx.serialization.json.internal/JsonStringBuilder.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package kotlinx.serialization.json.internal
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply formats/json/nativeMain/src/kotlinx.serialization.json.internal/JsonStringBuilder.kt:1]: Have no idea (pun intended) how you've found it

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt (hunk 1 above)
Comment (qwwdfsad) [reply formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:60]: Added tests, changed code layout, verified our benchmarks one more time

File: formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt (hunk 2 above)
Comment (qwwdfsad) [reply formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt:83]: >Maybe we can make sure that at least one char array always stays in pool

It potentially can lead to a tremendous memory footprint of up to 4GB, this is not really desirable.

Even if JSON encoding allocates char arrays for large JSON, it's still okay, what else application author could have expected? We do not guarantee zero allocation anyway. I'd prefer keeping the pool an implementation detail. Ones caring about performance will figure out about pool themselves

File: formats/json/nativeMain/src/kotlinx.serialization.json.internal/JsonStringBuilder.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (sandwwraith) [reply formats/json/nativeMain/src/kotlinx.serialization.json.internal/JsonStringBuilder.kt:1]: It's visible in file names in Github

**Review by sandwwraith — APPROVED** [review]

**Comment by ankushg** [comment]
> I suppose it's not really clear where all these improvements came from, so feel free to ask for additional comments, meanwhile, I'll try to document all my finding separately

These perf wins look great!

Out of curiosity, how many of the JSON<>String perf wins from 1.2 also apply to JSON<>`dynamic`?

**Comment by sandwwraith** [comment]
@ankushg No, dynamic conversion is a different code part
